### ISSUE TITLE
TestDropAllRace is failing on Teamcity with data race

### ISSUE THREAD

**Issue by ashish-goswami** [dgraph-io/badger#793]
Line to build https://teamcity.dgraph.io/viewLog.html?buildId=11071&buildTypeId=Badger_UnitTests&tab=buildLog
Stack trace attached.
```
[13:34:02][Step 2/2] WARNING: DATA RACE
[13:34:02][Step 2/2] Write at 0x00c0007a2190 by goroutine 209:
[13:34:02][Step 2/2]   runtime.closechan()
[13:34:02][Step 2/2]       /usr/local/go/src/runtime/chan.go:327 +0x0
[13:34:02][Step 2/2]   github.com/dgraph-io/badger.(*DB).doWrites()
[13:34:02][Step 2/2]       /home/pawan0201/go/src/github.com/dgraph-io/badger/db.go:710 +0x6b1
[13:34:02][Step 2/2] 
[13:34:02][Step 2/2] Previous read at 0x00c0007a2190 by goroutine 275:
[13:34:02][Step 2/2]   runtime.chansend()
[13:34:02][Step 2/2]       /usr/local/go/src/runtime/chan.go:140 +0x0
[13:34:02][Step 2/2]   github.com/dgraph-io/badger.(*DB).sendToWriteCh()
[13:34:02][Step 2/2]       /home/pawan0201/go/src/github.com/dgraph-io/badger/db.go:660 +0x32a
[13:34:02][Step 2/2]   github.com/dgraph-io/badger.(*Txn).commitAndSend()
[13:34:02][Step 2/2]       /home/pawan0201/go/src/github.com/dgraph-io/badger/txn.go:547 +0x783
[13:34:02][Step 2/2]   github.com/dgraph-io/badger.(*Txn).CommitWith()
[13:34:02][Step 2/2]       /home/pawan0201/go/src/github.com/dgraph-io/badger/txn.go:651 +0x218
[13:34:02][Step 2/2]   github.com/dgraph-io/badger.(*Txn).CommitAt()
[13:34:02][Step 2/2]       /home/pawan0201/go/src/github.com/dgraph-io/badger/managed_db.go:56 +0xa8
[13:34:02][Step 2/2]   github.com/dgraph-io/badger.TestDropAllRace.func1()
[13:34:02][Step 2/2]       /home/pawan0201/go/src/github.com/dgraph-io/badger/managed_db_test.go:317 +0x3a2
[13:34:02][Step 2/2] 
[13:34:02][Step 2/2] Goroutine 209 (running) created at:
[13:34:02][Step 2/2]   github.com/dgraph-io/badger.Open()
[13:34:02][Step 2/2]       /home/pawan0201/go/src/github.com/dgraph-io/badger/db.go:316 +0x1211
[13:34:02][Step 2/2]   github.com/dgraph-io/badger.TestDropAllRace()
[13:34:02][Step 2/2]       /home/pawan0201/go/src/github.com/dgraph-io/badger/managed_db_test.go:298 +0x2ca
[13:34:02][Step 2/2]   testing.tRunner()
[13:34:02][Step 2/2]       /usr/local/go/src/testing/testing.go:827 +0x162
[13:34:02][Step 2/2] 
[13:34:02][Step 2/2] Goroutine 275 (running) created at:
[13:34:02][Step 2/2]   github.com/dgraph-io/badger.TestDropAllRace()
[13:34:02][Step 2/2]       /home/pawan0201/go/src/github.com/dgraph-io/badger/managed_db_test.go:304 +0x37c
[13:34:02][Step 2/2]   testing.tRunner()
[13:34:02][Step 2/2]       /usr/local/go/src/testing/testing.go:827 +0x162
[13:34:02][Step 2/2] ==================
```

**Comment by jarifibrahim** [comment]
`DropAllPrefixRace` failed with a similar error multiple times.

**Comment by campoy** [comment]
Hey @jarifibrahim, this seems like a potentially serious issue.

Were you able to reproduce? And if so, do we have any plans on how to fix it?

**Comment by jarifibrahim** [comment]
Drop Prefix race failure with data race (https://teamcity.dgraph.io/downloadBuildLog.html?buildId=12993&plain=true) and deadlock (https://teamcity.dgraph.io/downloadBuildLog.html?buildId=12955&plain=true)

**Comment by campoy** [comment]
I totally forgot about this issue.

Is anyone working on it? @danielmai or @jarifibrahim, maybe?

Adding it to the v2.0.0 milestone just in case this is something serious.

**Comment by danielmai** [comment]
These three issues are still happening on master (d562bfa6aa1fec24f377dcc72f069a96cca95a8b).

They don't happen all the time. I see them after running the tests continuously after a few minutes.

```
while go test -v -race -run='TestDropAllRace' -timeout=1m; do :; done
```

```
while go test -v -race -run='TestDropPrefixRace' -timeout=1m; do :; done
```

**Comment by campoy** [comment]
Do they also occur in previous versions (e.g 1.5.5) or is it something we introduced recently?

**Comment by danielmai** [comment]
The TestDropAllRace data race does not happen in v1.5.5 (and DropPrefix isn't in v1.5.5).

**Comment by danielmai** [comment]
This is what the race detector found for TestDropAllPrefix.
```
WARNING: DATA RACE
Write at 0x00c0004c23d0 by goroutine 77:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:334 +0x0
  github.com/dgraph-io/badger.(*DB).doWrites()
      /home/dmai/go/src/github.com/dgraph-io/badger/db.go:731 +0x65f

Previous read at 0x00c0004c23d0 by goroutine 80:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:142 +0x0
  github.com/dgraph-io/badger.(*DB).sendToWriteCh()
      /home/dmai/go/src/github.com/dgraph-io/badger/db.go:681 +0x2f8
  github.com/dgraph-io/badger.(*Txn).commitAndSend()
      /home/dmai/go/src/github.com/dgraph-io/badger/txn.go:495 +0x748
  github.com/dgraph-io/badger.(*Txn).CommitWith()
      /home/dmai/go/src/github.com/dgraph-io/badger/txn.go:599 +0x21a
  github.com/dgraph-io/badger.(*Txn).CommitAt()
      /home/dmai/go/src/github.com/dgraph-io/badger/managed_db.go:56 +0xac
  github.com/dgraph-io/badger.TestDropAllRace.func1()
      /home/dmai/go/src/github.com/dgraph-io/badger/managed_db_test.go:317 +0x539

Goroutine 77 (running) created at:
  github.com/dgraph-io/badger.Open()
      /home/dmai/go/src/github.com/dgraph-io/badger/db.go:332 +0x19f9
  github.com/dgraph-io/badger.TestDropAllRace()
      /home/dmai/go/src/github.com/dgraph-io/badger/managed_db_test.go:298 +0x162
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:865 +0x163

Goroutine 80 (running) created at:
  github.com/dgraph-io/badger.TestDropAllRace()
      /home/dmai/go/src/github.com/dgraph-io/badger/managed_db_test.go:304 +0x2c2
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:865 +0x163
==================
```

**Comment by campoy** [comment]
Hey @ashish-goswami,

Could you please fix this? According to @manishrjain this should help:

```diff
diff --git a/db.go b/db.go
index 21bb22d..77187f0 100644
--- a/db.go
+++ b/db.go
@@ -658,6 +658,9 @@ func (db *DB) writeRequests(reqs []*request) error {
 }
 
 func (db *DB) sendToWriteCh(entries []*Entry) (*request, error) {
+    // Acquire mutex lock here.
+    sendLock.Lock()
+    defer sendLock.Unlock()
     if atomic.LoadInt32(&db.blockWrites) == 1 {
         return nil, ErrBlockedWrites
     }
@@ -1301,6 +1304,8 @@ func (db *DB) prepareToDrop() func() {
     }
     // Stop accepting new writes.
     atomic.StoreInt32(&db.blockWrites, 1)
+    sendLock.Lock()
+    defer sendLock.Unlock()
 
     // Make all pending writes finish. The following will also close writeCh.
     db.closers.writes.SignalAndWait()
```

**Comment by ashish-goswami** [comment]
@campoy Sure. I will check this today.

**Comment by campoy** [comment]
I think @jarifibrahim is already on it with #915, so coordinate with him

**Comment by jarifibrahim** [comment]
`DropAllRace` and `DropPrefixRace` have been failing because of different issues. `DropPrefixRace` fails because of https://github.com/dgraph-io/badger/issues/916.

 `TestDropAllRace` should be fixed by https://github.com/dgraph-io/badger/pull/915

**Comment by campoy** [comment]
Closing this issue, since the remaining fix is being tracked by #916
